### PR TITLE
PLAT-2322: Specify worker-port-list only if worker ports are defined

### PR DESCRIPTION
**Pull request by pocheung1**
### What problem does this PR solve?
To enable Istio for Ray, the Distributed Compute Operator replaces random worker ports with static ones and defaults to 5 worker ports.  This means that only up to 5 worker processes are allowed per worker pod.  This causes Ray programs to fail if they try to schedule more than 5 worker processes based on the resource requirements and the amount of cpu/gpu available in the hardware tiers.

https://github.com/cerebrotech/domino/pull/24018 addresses this by configuring the worker port list to 100 ports by default (worker port range configurable in Central Config) across all deployments.

For non-Istio deployments, we should not specify any static worker ports at all (which is only required for Istio deployments). 

### What is the solution?
- Nucleus will configure `spec.workerPorts` for Istio deployments only
- Remove default worker ports
- Specify worker-port-list only if worker ports are defined

https://dominodatalab.atlassian.net/browse/PLAT-2322

### CODE REVIEW

**Comment by pocheung1** [comment]
Verified that `--worker-port-list` is set only for Istio deployment.

Regular Ray worker pod:
```
spec:
  containers:
  - name: ray
    args:
    - start
    - --block
    - --node-ip-address=$(MY_POD_IP)
    - --num-cpus=$(MY_CPU_REQUEST)
    - --object-manager-port=2384
    - --node-manager-port=2385
    - --head
    - --ray-client-server-port=10001
    - --port=6379
    - --redis-shard-ports=6380,6381
    - --gcs-server-port=2386
    - --include-dashboard=true
    - --dashboard-host=0.0.0.0
    - --dashboard-port=8265
    command:
    - ray
```

Istio Ray worker pod:
```
spec:
  containers:
  - name: ray
    args:
    - start
    - --block
    - --node-ip-address=$(MY_POD_IP)
    - --num-cpus=$(MY_CPU_REQUEST)
    - --object-manager-port=2384
    - --node-manager-port=2385
    - --worker-port-list=11000,11001,11002,11003,11004,11005,11006,11007,11008,11009,11010,11011,11012,11013,11014,11015,11016,11017,11018,11019,11020,11021,11022,11023,11024,11025,11026,11027,11028,11029,11030,11031,11032,11033,11034,11035,11036,11037,11038,11039,11040,11041,11042,11043,11044,11045,11046,11047,11048,11049,11050,11051,11052,11053,11054,11055,11056,11057,11058,11059,11060,11061,11062,11063,11064,11065,11066,11067,11068,11069,11070,11071,11072,11073,11074,11075,11076,11077,11078,11079,11080,11081,11082,11083,11084,11085,11086,11087,11088,11089,11090,11091,11092,11093,11094,11095,11096,11097,11098,11099
    - --head
    - --ray-client-server-port=10001
    - --port=6379
    - --redis-shard-ports=6380,6381
    - --gcs-server-port=2386
    - --include-dashboard=true
    - --dashboard-host=0.0.0.0
    - --dashboard-port=8265
    command:
    - ray
```

File: pkg/resources/ray/statefulset.go [hunk 1]
```diff
@@ -318,7 +318,10 @@ func processArgs(rc *dcv1alpha1.RayCluster) []string {
 		"--num-cpus=$(MY_CPU_REQUEST)",
 		fmt.Sprintf("--object-manager-port=%d", rc.Spec.ObjectManagerPort),
 		fmt.Sprintf("--node-manager-port=%d", rc.Spec.NodeManagerPort),
-		fmt.Sprintf("--worker-port-list=%s", strings.Join(util.IntsToStrings(rc.Spec.WorkerPorts), ",")),
+	}
+
+	if rc.Spec.WorkerPorts != nil {
+		args = append(args, fmt.Sprintf("--worker-port-list=%s", strings.Join(util.IntsToStrings(rc.Spec.WorkerPorts), ",")))
```
Comment (fernandoacorreia) [inline pkg/resources/ray/statefulset.go:324]: Is there a test for this new behavior?

Comment (pocheung1) [reply pkg/resources/ray/statefulset.go:324]: There is an existing test with worker ports set.  I didn't add a new test for empty worker ports given my limited ability in Go.  I will defer that to Sonny after he is back from vacation.

**Review by saahildhulla — APPROVED** [review]
looks good to me, thanks Po!

File: pkg/resources/ray/statefulset.go (hunk 1 above)
Comment (fernandoacorreia) [reply pkg/resources/ray/statefulset.go:324]: @sonnysideup I recommend adding a new test case for this.

Notice that this PR decreased test coverage.

Comment (pocheung1) [reply pkg/resources/ray/statefulset.go:324]: The test coverage report looks iffy.

The report included 2 files that this PR did not modify where test coverage decreased:
- controllers/raycluster_controller.go (-2.47%)
- controllers/sparkcluster_controller.go (-1.34%)

A few lines got removed in this file and the test coverage decreased slightly, which is understandable:
- api/v1alpha1/raycluster_webhook.go (-0.10%)

A couple of new lines got added to this file and the test coverage remained at 100%:
- pkg/resources/ray/statefulset.go

Overall there is no net decrease in the number of lines covered in this PR.
